### PR TITLE
Jseanxu/accents

### PR DESCRIPTION
**Pull request by jseanxu**

### CODE REVIEW

File: src/js/WinJS/Accents.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved. Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Accents should start with an underscore because it is private

File: src/js/ui.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation.  All Rights Reserved. Licensed under the MIT License. See License.txt in the project root for license information.
 define([
     'WinJS/Core/_WinJS',
+    'WinJS/Accents',
```
Comment (rigdern) [inline src/js/ui.js]: Delete this. Controls which depend on it will cause it to get pulled in.

File: src/js/WinJS/Core/_WinRT.d.ts [hunk 3]
```diff
@@ -181,7 +181,27 @@ export declare module Windows {
 
 export declare module Windows {
     export module UI {
+        export class Color {
```
Comment (rigdern) [inline src/js/WinJS/Core/_WinRT.d.ts]: How will we contribute these back to defintely typed?

File: src/js/WinJS/Accents.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+_BaseUtils.ready().then(() => {
+    // Figure out color theme
+    var tag = _Global.document.createElement("winjs");
+    _Global.document.body.appendChild(tag);
+    var theme = _Global.getComputedStyle(tag).opacity;
+    isDarkTheme = theme === "0";
+    tag.parentElement.removeChild(tag);
+
+    if (_BaseUtils.hasWinRT) {
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Should feature detect like this: `if (_WinRT.Windows.UI.ViewManagement.UISettings)`

File: src/js/WinJS/Accents.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        UISettings = new _WinRT.Windows.UI.ViewManagement.UISettings();
+        UISettings.addEventListener("colorvalueschanged", handleColorsChanged);
+        handleColorsChanged();
+    } else {
+        // No WinRT - use hardcoded blue accent color
+        colors.push(
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Can you comment that these align with the ColorType enum?

File: src/js/WinJS/Accents.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    var UIColorType = _WinRT.Windows.UI.ViewManagement.UIColorType;
+    var accent = colorToString(UISettings.getColorValue(_WinRT.Windows.UI.ViewManagement.UIColorType.accent), 1);
+    if (colors[0] === accent) {
+        return;
+    }
+
+    // Establish colors
+    colors = [];
+    colors.push(accent);
+    colors.push(
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Can you comment that these align with the ColorType enum?

File: src/js/WinJS/Accents.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    listSelectPressInverse = 6,
+}
+
+export function createAccentRule(selector: string, props: { name: string; value: ColorTypes; }[], noHoverSelector = false) {
+    rules.push({ selector: selector, props: props, noHoverSelector: noHoverSelector });
+    scheduleWriteRules();
+}
+
+function scheduleWriteRules() {
+    if (!rules.length || writeRulesTOHandle !== -1) {
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: === 0

File: src/js/WinJS/Accents.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function _reset() {
+    rules = [];
+    cleanup();
+}
+
+_BaseUtils.ready().then(() => {
+    // Figure out color theme
+    var tag = _Global.document.createElement("winjs");
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: themeDetectionTag

File: src/js/WinJS/Accents.ts [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            // example rule: { selector: "  .foo,  .bar:hover ,  div:hover  ", props: [{ name: "color", value: 0 }, { name: "background-color", value: 1 }, noHoverSelector: false }
+
+            var body = "  " + rule.props.map(prop => prop.name + ": " + colors[prop.value] + ";").join("\n  ");
+            // body = color: *accent*; background-color: *listSelectHover*
+
+            var selectorSplit = rule.selector.split(",").map(str => str.trim()); // [".foo", ".bar:hover", "div"]
+            var selector = selectorSplit.join(",\n"); // ".foo, .bar:hover, div"
+
+            // Hover Selectors
+            if (!rule.noHoverSelector) {
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Remove. Have developer specify `html.win-hoverable` manually.

Make sure to handle inverse theme correctly with `html.win-hoverable` (i.e. `win-ui-light` class has to come after the `html.win-overable` and not before)

File: src/js/WinJS/Accents.ts [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+var isDarkTheme = false;
+var rules: { selector: string; props: { name: string; value: ColorTypes; }[]; noHoverSelector: boolean; }[] = [];
+var writeRulesTOHandle = -1;
+
+export enum ColorTypes {
+    accent = 0,
+    listSelectRest = 1,
+    listSelectHover = 2,
+    listSelectPress = 3,
+    listSelectRestInverse = 4,
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Can you underscore the 3 inverse color because they are private?

File: src/js/WinJS/Accents.ts [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            var css = selector + " {\n" + body + "\n}";
+            // css = .foo, .bar:hover, div:hover, html.win-hoverable .bar:hover, html.win-hoverable.bar:hover {body}
+
+            // Inverse Theme Selectors
+            var isThemedColor = rule.props.some(prop => prop.value !== ColorTypes.accent)
+            if (isThemedColor) {
+                var inverseBody = "  " + rule.props.map(prop => prop.name + ": " + colors[(prop.value ? (prop.value + 3) : prop.value)] + ";").join("\n  ");
+                // inverseBody = "color: *accent*; background-color: *listSelectHoverInverse"
+
+                selectorSplit.forEach(sel => {
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: hover selectors aren't included here

File: src/js/WinJS/Accents.ts [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                    css += "\n" + inverseThemeSelector + " " + sel;
+                    if (CSSSelectorTokens.indexOf(sel[0]) !== -1) {
+                        css += ",\n" + inverseThemeSelector + sel;
+                    }
+                    css += " {\n" + inverseBody + "\n}";
+                });
+                // css
+                //.foo, .bar:hover, div:hover, html.win-hoverable .bar:hover, html.win-hoverable.bar:hover {body} 
+                //.win-ui-light .foo {inverseBody}
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: comment should only include one instance of inverseBody

File: src/js/WinJS/Accents.ts [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+function handleColorsChanged() {
+    var UIColorType = _WinRT.Windows.UI.ViewManagement.UIColorType;
+    var accent = colorToString(UISettings.getColorValue(_WinRT.Windows.UI.ViewManagement.UIColorType.accent), 1);
+    if (colors[0] === accent) {
+        return;
+    }
+
+    // Establish colors
+    colors = [];
```
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: colors.length = 0 so colors in toPublish will always refer to the latest copy

File: tests/TestLib/winjs.dev.d.ts [hunk 14]
```diff
@@ -162,6 +162,23 @@ declare module WinJS {
             function _clickEaterTapped(): void;
         }
 
+        module _Accents {
+            export enum ColorTypes {
+                accent,
+                listSelectRest,
+                listSelectHover,
+                listSelectPress,
+                listSelectRestInverse,
```
Comment (rigdern) [inline tests/TestLib/winjs.dev.d.ts]: Make the 3 inverse colors private

File: tests/UI/accents.ts [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.listSelectHoverInverse], cs.columnRuleColor);
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.listSelectPressInverse], cs.outlineColor);
+
+            Accents.createAccentRule("#accent-test-element", [
+                { name: "background-color", value: Accents.ColorTypes.listSelectRest },
+                { name: "column-rule-color", value: Accents.ColorTypes.listSelectHover },
+                { name: "outline-color", value: Accents.ColorTypes.listSelectPress },
+            ]);
+
+            Accents.createAccentRule("#accent-test-element", [{ name: "color", value: Accents.ColorTypes.listSelectHover }]);
```
Comment (rigdern) [inline tests/UI/accents.ts]: Is this line needed?

File: tests/UI/accents.ts [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+                complete();
+            });
+        }
+
+        testCreateAccentRuleInverseNoLeadingToken(complete) {
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.listSelectHover], getComputedStyle(testElement).color);
+
+            Accents.createAccentRule("randomTag", [{ name: "color", value: Accents.ColorTypes.listSelectHover }]);
+            WinJS.Promise.timeout().done(() => {
+                LiveUnit.Assert.isTrue(getDynamicStyleElement().textContent.indexOf("randomTag") !== -1);
```
Comment (rigdern) [inline tests/UI/accents.ts]: isTrue(x !== -1)
isFalse(x !== -1)

It's hard to tell which is which at a glance. And this is a common pattern. Perhaps you should have a couple of helpers:
  - assertContains
  - assertDoesntContain

File: tests/UI/accents.ts [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        testMultipleSelectors(complete) {
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.accent], getComputedStyle(testElement).color);
+
+            Accents.createAccentRule("#accent-test-element, #foo", [{ name: "color", value: Accents.ColorTypes.accent }]);
+            WinJS.Promise.timeout().done(() => {
+                LiveUnit.Assert.isTrue(getDynamicStyleElement().textContent.indexOf("#accent-test-element") !== -1);
+                LiveUnit.Assert.isTrue(getDynamicStyleElement().textContent.indexOf("#foo") !== -1);
+                LiveUnit.Assert.areEqual(Accents._colors[Accents.ColorTypes.accent], getComputedStyle(testElement).color);
```
Comment (rigdern) [inline tests/UI/accents.ts]: You should also have an element with id `foo` and test against its color

File: tests/UI/accents.ts [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+                complete();
+            });
+        }
+
+        testMultipleSelectorsWithWhitespace(complete) {
+            Accents.createAccentRule("   #accent-test-element     ,   #foo     ", [{ name: "color", value: Accents.ColorTypes.accent }]);
+            WinJS.Promise.timeout().done(() => {
+                LiveUnit.Assert.isTrue(getDynamicStyleElement().textContent.indexOf("#accent-test-element") !== -1);
+                LiveUnit.Assert.isTrue(getDynamicStyleElement().textContent.indexOf("#foo") !== -1);
+                LiveUnit.Assert.areEqual(Accents._colors[Accents.ColorTypes.accent], getComputedStyle(testElement).color);
```
Comment (rigdern) [inline tests/UI/accents.ts]: You should also have an element with id foo and test against its color

File: tests/UI/accents.ts [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.listSelectHoverInverse], cs.columnRuleColor);
+            LiveUnit.Assert.areNotEqual(Accents._colors[Accents.ColorTypes.listSelectPressInverse], cs.outlineColor);
+
+            Accents.createAccentRule("#accent-test-element, randomTag", [
+                { name: "background-color", value: Accents.ColorTypes.listSelectRest },
+                { name: "column-rule-color", value: Accents.ColorTypes.listSelectHover },
+                { name: "outline-color", value: Accents.ColorTypes.listSelectPress },
+            ]);
+
+            Accents.createAccentRule("randomTag, #accent-test-element", [{ name: "color", value: Accents.ColorTypes.listSelectHover }]);
```
Comment (rigdern) [inline tests/UI/accents.ts]: Is this rule unused?

File: src/js/WinJS/Accents.ts (hunk 1 above)
Comment (rigdern) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/ui.js (hunk 2 above)
Comment (jseanxu) [inline src/js/ui.js]: Done

File: src/js/WinJS/Accents.ts (hunk 4 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 7 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 12 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 13 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 6 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 8 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 5 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 10 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: tests/TestLib/winjs.dev.d.ts (hunk 14 above)
Comment (jseanxu) [inline tests/TestLib/winjs.dev.d.ts]: Done

File: tests/UI/accents.ts (hunk 15 above)
Comment (jseanxu) [inline tests/UI/accents.ts]: Removed

File: tests/UI/accents.ts (hunk 19 above)
Comment (jseanxu) [inline tests/UI/accents.ts]: Removed

File: tests/UI/accents.ts (hunk 17 above)
Comment (jseanxu) [inline tests/UI/accents.ts]: Done

File: tests/UI/accents.ts (hunk 18 above)
Comment (jseanxu) [inline tests/UI/accents.ts]: Done

File: tests/UI/accents.ts (hunk 16 above)
Comment (jseanxu) [inline tests/UI/accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 9 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Done

File: src/js/WinJS/Accents.ts (hunk 11 above)
Comment (jseanxu) [inline src/js/WinJS/Accents.ts]: Fixed

**Pull request closed without merging**